std/format.d

```diff
@@ -5771,18 +5771,45 @@ package static const checkFormatException(alias fmt, Args...) =
     return null;
 }();
 
-/*****************************************************
+/**
  * Format arguments into a string.
  *
+ * If the format string is fixed, passing it as a template parameter checks the
+ * type correctness of the parameters at compile-time. This also can result in
+ * better performance.
+ *
  * Params: fmt  = Format string. For detailed specification, see $(LREF formattedWrite).
  *         args = Variadic list of arguments to _format into returned string.
+ *
+ * Throws:
+ *     $(LREF, FormatException) if the number of arguments doesn't match the number
+ *     of format parameters and vice-versa.
  */
 typeof(fmt) format(alias fmt, Args...)(Args args)
 if (isSomeString!(typeof(fmt)))
 {
+    import std.array : appender;
+
     alias e = checkFormatException!(fmt, Args);
+    alias Char = Unqual!(ElementEncodingType!(typeof(fmt)));
+
     static assert(!e, e.msg);
-    return .format(fmt, args);
+    auto w = appender!(immutable(Char)[]);
+
+    // no need to traverse the string twice during compile time
+    if (!__ctfe)
+    {
+        enum len = guessLength!Char(fmt);
+        w.reserve(len);
+    }
+    else
+    {
+        enum len = fmt.length;
+        w.reserve(len);
```

Comment:
Hmm, just out of curiosity. Why not use `guessLength` at runtime too?  I think runtime performance could benefit from not having to reallocate as frequently.